Store the parameters of the chosen connection when using Master/Slave.

xrash opened the pull request:
When using the MasterSlaveConnection, the internal parameters array has a different structure than the usual Connection. This can cause the following methods to fail: getHost(), getPort(), getUsername() and getPassword().

I changed the MasterSlaveConnection so that it stores the parameters of the chosen connection in a private property. With this change, the methods that used to fail can now retrieve the configuration for the actual chosen connection, which is, in my opinion, clearly the expected behavior.

I chose to create a new private property in MasterSlaveConnection instead of inherit the $_params from Connection in order to not cause any confusion by having two different array structures in the same variable at different situations.

I also expanded the tests for this case.

xrash committed:
Store the parameters of the chosen connection when using the Master/Slave config.

xrash committed:
Add one more test case in the MasterSlaveConnectionTest.

Ocramius commented on tests/Doctrine/Tests/DBAL/Functional/MasterSlaveConnectionTest.php:
```diff
@@ -139,4 +143,34 @@ public function testMasterSlaveConnectionCloseAndReconnect()
+
+    public function testMasterSlaveConnectionParams()
+    {
+        $test = function($connectionName, $_this, $expected) {
```
Could you use `$that` instead of `$_this`? (just nitpicking)

Ocramius replied:
Also, the callback could be renamed for clarity

Ocramius commented on tests/Doctrine/Tests/DBAL/Functional/MasterSlaveConnectionTest.php:
```diff
@@ -139,4 +143,34 @@ public function testMasterSlaveConnectionCloseAndReconnect()
+            );
+
+            foreach ($expected as $key => $value) {
+                $_this->assertEquals($value, $params[$key]);
```
Should probably `assertSame` rather than `assertEquals`

Ocramius replied:
Can probably directly `assertSame` against the given array rather than iterating over `$expected`

xrash committed:
Improve the MasterSlaveConnectionTest.

Now the tests of the MasterSlaveConnection actually use different parameters
for different connections. The previous setup had a master and two slaves, but all
shared the same configuration. Now, the master uses the main configuration and
the two slaves share a different configuration. This enables more robust testing.

xrash committed:
Fix a subtle bug in MasterSlaveConnection::getDatabase().

The Connection::getDatabase() method can execute the query() method
in order to perform a SELECT DATABASE() query to get the current
database name.

The MasterSlaveConnection cannot reuse such a behavior because the
execution of the query() method will cause the instant connection
to the master database, defeating the purpose of the getDatabase()
method for any slave.

This change considers only the database name in the configuration,
ignoring any possible subsequent database change.

morozov commented:
Closing as stale.
